parsers/CA_ON.py

```diff
@@ -64,331 +55,265 @@
+            datetime=datetime.datetime.combine(date, time, tzinfo=TIMEZONE),
+            production=mix,
+            source=SOURCE,
+            zoneKey=ZONE_KEY,
         )
-    )
-
-    df = pd.DataFrame(all_productions)
-
-    # group individual plants using the same fuel together for each time period
-    by_fuel = df.groupby(["dt", "fuel"]).sum().unstack()
-    # to debug, you can `print(by_fuel)` here, which gives a very pretty table
-
-    by_fuel_dict = by_fuel["production"].to_dict("index")
 
-    data = [
-        {
-            "datetime": time.to_pydatetime(),
-            "zoneKey": zone_key,
-            "production": productions,
-            "storage": {},
-            "source": "ieso.ca",
-        }
-        for time, productions in by_fuel_dict.items()
-    ]
-
-    # being constructed from a dict, data is not guaranteed to be in chronological order.
-    # sort it for clean-ness and easier debugging.
-    data = sorted(data, key=lambda dp: dp["datetime"])
-
-    return data
+    return production_breakdowns.to_list()
 
 
 def fetch_price(
-    zone_key: str = "CA-ON",
+    zone_key: str = ZoneKey("CA-ON"),
     session: Session | None = None,
-    target_datetime: datetime | None = None,
+    target_datetime: datetime.datetime | None = None,
     logger: Logger = getLogger(__name__),
-) -> list[dict]:
+) -> list[dict[str, Any]]:
     """Requests the last known power price per MWh of a given region."""
 
-    # "HOEP" below is "Hourly Ontario Energy Price".
-    # There also exists a 5-minute price, but its archives only go back roughly 4 days
-    # (see http://www.ieso.ca/power-data "5 Minute Market Clearing Price").
+    if zone_key != ZONE_KEY:
+        raise NotImplementedError(f"unimplemented zone '{zone_key}'")
 
-    dt, xml = _fetch_ieso_xml(target_datetime, session, logger, PRICE_URL)
+    date, xml = _fetch_xml(logger, session, target_datetime, PRICE_URL)
 
     if not xml:
         return []
 
-    prices = (
-        xml.find(XML_NS_TEXT + "IMODocBody")
-        .find(XML_NS_TEXT + "HOEPs")
-        .findall(XML_NS_TEXT + "HOEP")
-    )
-
-    data = [
-        {
-            "datetime": _parse_ieso_hour(price, dt),
-            "price": float(price.find(XML_NS_TEXT + "Price").text),
-            "currency": "CAD",
-            "source": "ieso.ca",
-            "zoneKey": zone_key,
-        }
-        for price in prices
-    ]
+    # "HOEP" stands for "Hourly Ontario Energy Price". There also exists a
+    # 5-minute price, but its archives only go back roughly 4 days (see "5
+    # Minute Market Clearing Price" at http://www.ieso.ca/power-data ).
+    prices = PriceList(logger)
+    for hoep in xml.iter(NAMESPACE + "HOEP"):
+        prices.append(
+            currency="CAD",
+            datetime=datetime.datetime.combine(
+                date, datetime.time(hour=_parse_hour(hoep)), tzinfo=TIMEZONE
+            ),
+            price=float(hoep.findtext(NAMESPACE + "Price")),
```

Comment:
While the price argument handles None values the float does not.

I would assign the value to a variable and then check so that variable is not None.
Something like:
```python
price_string = hoep.findtext(NAMESPACE + "Price")
price = float(price_string) if price_string else None
```